c++/src/kj/compat/url.c++

```diff
@@ -40,27 +40,27 @@ constexpr auto END_AUTHORITY = parse::anyOfChars("/?#");
 // not currently allow it in the authority component, because our parser would reject it as a host
 // character anyway.
 
-const parse::CharGroup_& getEndPathPart(Url::Context context) {
+const parse::CharGroup_& getEndPathPart(Url::Options::MergeFragment mergeFragment) {
   static constexpr auto END_PATH_PART_HREF = parse::anyOfChars("/?#");
   static constexpr auto END_PATH_PART_REQUEST = parse::anyOfChars("/?");
 
-  switch (context) {
-    case Url::REMOTE_HREF:        return END_PATH_PART_HREF;
-    case Url::HTTP_PROXY_REQUEST: return END_PATH_PART_REQUEST;
-    case Url::HTTP_REQUEST:       return END_PATH_PART_REQUEST;
+  switch (mergeFragment) {
+    case Url::Options::MergeFragment::AUTO:  KJ_UNREACHABLE;
```

Comment:
It's not immediately clear to me this is actually unreachable, especially since the default value for the variable being supplied to this function is `AUTO`. Should this perhaps be a `KJ_REQUIRE` so that we generate an exception instead of undefined behavior?

Perhaps you're already aware (but in case you're not), `KJ_UNREACHABLE` expands to `__builtin_unreachable` which is a signal to the compiler that hitting this code path is impossible & the compiler can treat it as undefined behavior (aka UB) for optimization purposes (i.e. colloquially one should think of hitting a UB statement as potentially running `rm -rf /`).

For example, this is generally fine when outside of a an exhaustive `switch` statement (as seen below) over a closed enum (i.e. generally what enums internal to a program represent). The reason for this is C/C++ define enums as open & thus that knowledge isn't available to the compiler to understand your intention (hence the builtin). Arguably, this could have been fixed by making `enum class` closed by default or something but alas we're stuck with `__builtin_unreachable`.

Alternatively, you could make the `mergeFragment` variable a `Maybe` & get rid of the `AUTO` state. That would remove the `AUTO` state from switch statements (& thus ensure you did `KJ_REQUIRE_NONNULL` call when unwrapping).

Reply:
This is interesting. In the form I have now, you are correct, it is reachable if someone invokes `tryParseWithoutContext()`. That said, if I do your suggestion with `Maybe<Context>` in the public signature, this can become truly unreachable. I'd rather `KJ_FAIL_REQUIRE` here I think, that way I can explain why this should never happen.

I was a little surprised to read that it uses a `__builtin` since those usually aren't cross-platform. Traditionally, I've seen either an uncatchable exception thrown or an abort instead.